# 🧼 Data Cleaning Documentation

In [160]:
import pandas as pd
import re
from sklearn.feature_extraction.text import CountVectorizer

In [32]:
file_path = "../data/freelancer_data.csv"  
df = pd.read_csv(file_path)

df.head()

name price  \
0  OweBest Technologies Pvt.   $18   
1     Elite Information Tech   $15   
2        TAFSOL Technologies   $45   
3               eTranslators   $30   
4  ✈THE MARKETING MAVERICKS™   $45   

                                               image     country  \
0  https://www.f-cdn.com/assets/img/membership/ba...       India   
1  https://www.f-cdn.com/assets/img/membership/ba...       India   
2  https://www.f-cdn.com/assets/img/membership/ba...    Pakistan   
3  https://www.f-cdn.com/assets/img/membership/ba...  Bangladesh   
4  https://www.f-cdn.com/assets/img/membership/ba...       India   

                                         description  \
0  OweBest is a dynamic and rapidly expanding IT ...   
1  7000 ⭐⭐⭐⭐⭐+ Reviews, 2000+ satisfied clients G...   
2  TAFSOL TECHNOLOGIES Group!! Aiming to deliver ...   
3  eTranslators offers over a DECADE of experienc...   
4  The Grace Of Almighty God ____________________...   

                                              skills  rating reviews  
0    PHPWebsite DesignHTMLMySQLSoftware Architecture     4.9     490  
1  SEOLink BuildingInternet MarketingAdvertisingG...     4.9    7018  
2  PHPWebsite DesignHTMLSoftware ArchitectureGrap...     5.0     195  
3  TranslationEnglish (US) TranslatorEnglish (UK)...     4.9    3633  
4  Internet MarketingFacebook MarketingSocial Med...     5.0     558

In [33]:
null_counts = df.isnull().sum()
null_counts

name              1
price            20
image          5146
country           1
description      13
skills           33
rating         4063
reviews           0
dtype: int64



## 📌 Why Not Remove NULLs from `image` and `rating`?

During the data preprocessing, we found that `image` and `rating` contain a significant number of NULL values:  

| Column       | NULL Count |
|-------------|-----------|
| `name`      | 1         |
| `price`     | 20        |
| `image`     | 5146      |
| `country`   | 1         |
| `description` | 13      |
| `skills`    | 33        |
| `rating`    | 4063      |
| `reviews`   | 0         |

Instead of **removing** these rows, we used a **data cleaning approach** to preserve as much useful information as possible.




In [34]:
unique_ratings = df['rating'].dropna().unique()
unique_ratings

array([4.9, 5. , 4.8, 4.6, 4.7, 4.4, 4.5, 4.2, 2.7, 4. , 3.8, 2.8, 3.6,
       3.4, 0. , 4.1, 3.7, 3.9, 4.3, 3.1, 3. , 3.3, 2.9, 1.4])

## ✅ **Handling NULL Values**
### 🔹 **`rating` (Freelancer Rating)**
**Problem:** Over 4,000 missing ratings.  
**Solution:**  
- **Replaced NULL values** with `"Not Rated"` instead of `0.0`.  
- This avoids misleading data while indicating that the freelancer has not been rated.

In [35]:
df['rating'].fillna("not rating", inplace=True)

C:\Users\YOUSSEF\AppData\Local\Temp\ipykernel_7864\1282182483.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['rating'].fillna("not rating", inplace=True)
C:\Users\YOUSSEF\AppData\Local\Temp\ipykernel_7864\1282182483.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'not rating' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df['rating'].fillna("not rating", inplace=True)


In [36]:
df['image'].fillna("no level", inplace=True)

C:\Users\YOUSSEF\AppData\Local\Temp\ipykernel_7864\1206456235.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['image'].fillna("no level", inplace=True)


In [37]:
null_counts = df.isnull().sum()
null_counts

name            1
price          20
image           0
country         1
description    13
skills         33
rating          0
reviews         0
dtype: int64

In [38]:
df.dropna(inplace=True)


In [39]:
null_counts = df.isnull().sum()
null_counts

name           0
price          0
image          0
country        0
description    0
skills         0
rating         0
reviews        0
dtype: int64

In [ ]:
unique_country = df['country'].dropna().unique()
unique_country

array(['India', 'Pakistan', 'Bangladesh', 'Croatia', 'Canada', 'Germany',
       'Armenia', 'United States', 'United Kingdom', 'Turkey', 'Greece',
       'Philippines', 'Romania', 'Vietnam', 'Sweden', 'Ukraine',
       'Sri Lanka', 'Mexico', 'Australia', 'Egypt',
       'United Arab Emirates', 'Indonesia', 'Kenya', 'Peru', 'Georgia',
       'Netherlands', 'Ethiopia', 'Bulgaria', 'Venezuela', 'Mauritius',
       'Nepal', 'Argentina', 'Jordan', 'Macedonia', 'Chile', 'Serbia',
       'China', 'Nigeria', 'Hong Kong', 'Colombia', 'Brazil', 'Belarus',
       'Moldova, Republic of', 'Czech Republic', 'Malaysia', 'Morocco',
       'Ireland', 'Bosnia and Herzegovina', 'Jamaica', 'Latvia',
       "Cote D'Ivoire", 'Ghana', 'Nicaragua', 'Cyprus', 'Austria',
       'Tanzania, United Republic of', 'Italy', 'Kazakhstan', 'Portugal',
       'Spain', 'Algeria', 'New Zealand', 'Korea, Republic of',
       'South Africa', 'Dominican Republic', 'Costa Rica', 'France',
       'Guatemala', 'Belgium', 'Thail

In [41]:
standardized_countries = {
    "Moldova, Republic of": "Moldova",
    "Korea, Republic of": "South Korea",
    "Tanzania, United Republic of": "Tanzania",
    "Cote D'Ivoire": "Ivory Coast",
    "Brunei Darussalam": "Brunei",
    "Sao Tome and Principe": "Sao Tome",
    "Palestinian Territory": "Palestine"
}
df['country'] = df['country'].replace(standardized_countries)


In [42]:
unique_country = df['country'].dropna().unique()
unique_country

array(['India', 'Pakistan', 'Bangladesh', 'Croatia', 'Canada', 'Germany',
       'Armenia', 'United States', 'United Kingdom', 'Turkey', 'Greece',
       'Philippines', 'Romania', 'Vietnam', 'Sweden', 'Ukraine',
       'Sri Lanka', 'Mexico', 'Australia', 'Egypt',
       'United Arab Emirates', 'Indonesia', 'Kenya', 'Peru', 'Georgia',
       'Netherlands', 'Ethiopia', 'Bulgaria', 'Venezuela', 'Mauritius',
       'Nepal', 'Argentina', 'Jordan', 'Macedonia', 'Chile', 'Serbia',
       'China', 'Nigeria', 'Hong Kong', 'Colombia', 'Brazil', 'Belarus',
       'Moldova', 'Czech Republic', 'Malaysia', 'Morocco', 'Ireland',
       'Bosnia and Herzegovina', 'Jamaica', 'Latvia', 'Ivory Coast',
       'Ghana', 'Nicaragua', 'Cyprus', 'Austria', 'Tanzania', 'Italy',
       'Kazakhstan', 'Portugal', 'Spain', 'Algeria', 'New Zealand',
       'South Korea', 'South Africa', 'Dominican Republic', 'Costa Rica',
       'France', 'Guatemala', 'Belgium', 'Thailand', 'Saudi Arabia',
       'Montenegro', 'Boliv

In [46]:
geo_classification = {
    "Asia": ['India', 'Pakistan', 'Bangladesh', 'Armenia', 'Turkey', 'Philippines', 'Vietnam', 'Sri Lanka',
             'Indonesia', 'Nepal', 'Georgia', 'Kazakhstan', 'Malaysia', 'China', 'Hong Kong', 'Japan',
             'South Korea', 'Uzbekistan', 'Bahrain', 'Oman', 'Maldives', 'Singapore', 'Israel',
             'Azerbaijan', 'Kuwait', 'Iraq', 'Yemen', 'Taiwan', 'Mongolia', 'Cambodia', 'Afghanistan', 'Kyrgyzstan', 'Tajikistan'],

    "Middle East & North Africa": ['Egypt', 'United Arab Emirates', 'Jordan', 'Saudi Arabia', 'Morocco',
                                    'Algeria', 'Tunisia', 'Lebanon', 'Palestine', 'Bahrain', 'Oman', 'Kuwait', 'Iraq', 'Yemen', 'Qatar'],

    "Europe": ['Croatia', 'Germany', 'United Kingdom', 'Greece', 'Romania', 'Sweden', 'Ukraine',
               'Netherlands', 'Bulgaria', 'Macedonia', 'Serbia', 'Belarus', 'Moldova',
               'Czech Republic', 'Ireland', 'Bosnia and Herzegovina', 'Latvia', 'Cyprus', 'Austria',
               'Italy', 'Portugal', 'Spain', 'Poland', 'Estonia', 'Slovakia', 'Hungary', 'Switzerland',
               'Lithuania', 'Malta', 'Kosovo', 'Luxembourg', 'Albania', 'Denmark', 'Montenegro'],

    "Africa": ['Kenya', 'Ethiopia', 'Ivory Coast', 'Ghana', 'Benin', 'Madagascar', 'Rwanda', 'Sierra Leone',
               'Angola', 'Cameroon', 'Mozambique', 'Zambia', 'South Africa', 'Tanzania', 'Zimbabwe', 'Mauritius', 'Sao Tome'],

    "North America": ['United States', 'Canada', 'Mexico', 'Jamaica', 'Trinidad and Tobago', 'Haiti',
                      'Dominican Republic', 'Puerto Rico', 'Nicaragua', 'El Salvador', 'Guatemala', 'Costa Rica', 'Panama', 'Honduras'],

    "South America": ['Peru', 'Venezuela', 'Argentina', 'Chile', 'Colombia', 'Brazil', 'Ecuador',
                      'Uruguay', 'Bolivia', 'Guyana'],

    "Oceania": ['Australia', 'New Zealand']
}
def classify_region(country):
    for region, countries in geo_classification.items():
        if country in countries:
            return region
    return "Other" 

In [47]:
df['Region'] = df['country'].apply(classify_region)

In [48]:
unique_Region = df['Region'].dropna().unique()
unique_Region

array(['Asia', 'Europe', 'North America', 'Oceania',
       'Middle East & North Africa', 'Africa', 'South America', 'Other'],
      dtype=object)

In [49]:
df['Region'].value_counts()

Region
Asia                          5999
North America                  792
Europe                         692
South America                  307
Middle East & North Africa     233
Africa                         182
Other                          109
Oceania                         87
Name: count, dtype: int64

# 🌍 Freelancer Region Analysis
Based on the dataset, **more than 50% of freelancers** come from **Asia**, making it the dominant region in the freelance market.

---

## 📊 **Freelancer Distribution by Region**

| Region                        | Number of Freelancers |
|--------------------------------|----------------------:|
| 🌏 **Asia**                   | **5,999** (🔹 **~63%** of total) |
| 🌎 North America              | 792   |
| 🌍 Europe                     | 692   |
| 🌎 South America              | 307   |
| 🌍 Middle East & North Africa | 233   |
| 🌍 Africa                     | 182   |
| 🌍 Other                      | 109   |
| 🌏 Oceania                    | 87    |

---

## ✅ **Key Insights**
1️⃣ **Asia Dominates** → With **~63% of all freelancers**, Asia leads the freelancing industry, likely due to a strong presence of tech talent in countries like **India, Pakistan, and Bangladesh**.  
2️⃣ **North America & Europe** → Contribute a **significant number** of freelancers but are much smaller compared to Asia.  
3️⃣ **Middle East & North Africa (MENA)** → A growing freelance market with **233 freelancers**.  
4️⃣ **Oceania & "Other" Regions** → Represent the **smallest share** of freelancers.  

---

In [71]:
unique_skills = df['skills'].dropna().unique()
len(unique_skills)

5662

In [64]:
def split_skills(text):
    return re.split(r' (?=[A-Z])', text) if pd.notna(text) else []

df['skills_list'] = df['skills'].apply(split_skills)

df['skills_list'][1]



['SEOLink', 'BuildingInternet', 'MarketingAdvertisingGoogle', 'Adwords']

In [65]:
df['skills_count'] = df['skills_list'].apply(len)

In [72]:
unique_skills_count = df['skills_count'].dropna().unique()
unique_skills_count

array([3, 4, 5, 2, 6, 1, 7, 8], dtype=int64)

# 🎯 Skill Range Analysis Across Freelancer Platforms  

## 📌 Overview  
Freelancers on different platforms have varying skill levels, ranging from **1 (Beginner) to 8 (Expert)**.  
The identified **skill range** is:  
## 📊 **Skill Levels and Interpretation**  

| **Skill Level** | **Description** |
|---------------:|------------------------------------------|
| **1-2**        | 🟢 Beginner: Basic understanding, minimal experience. |
| **3-4**        | 🔵 Intermediate: Some expertise, practical experience. |
| **5-6**        | 🟠 Advanced: Strong experience, specialized knowledge. |
| **7-8**        | 🔴 Expert: Highly skilled, industry leader. |

---

In [76]:
df['image']	[10]

'https://www.f-cdn.com/assets/img/membership/badges/a-membership-level-5-badge-5a58886e.svg'

In [77]:
def extract_level(text):
    match = re.search(r'level-\d+', text)
    return match.group() if match else "No Level"
df['level'] = df['image'].apply(extract_level)

In [78]:
unique_level = df['level'].dropna().unique()
unique_level

array(['level-5', 'level-3', 'level-4', 'level-2', 'No Level', 'level-1'],
      dtype=object)

In [79]:
df.drop(columns=['image'], inplace=True)
df.drop(columns=['skills'], inplace=True)

In [81]:
df['price'] = df['price'].replace('[^\d.]', '', regex=True)

In [85]:
df["skills_list"][0]

['PHPWebsite', 'DesignHTMLMySQLSoftware', 'Architecture']

In [90]:
unique_skill = list(set(sum(df["skills_list"], [])))
print(unique_skill)

['ManagementWeb', 'BlogBook', 'ArduinoCircuit', 'EntryEditingFlyer', 'XcodeiOS', 'CoPilotiPhone', 'WritingBlogClassifieds', 'SourcingEnglish (UK)', 'IntegrationBubble', 'Mechanical', 'TestingReport', 'AdwordsAdvertisingFacebook', 'DesignStationery', 'DesignUX /', '.NETActionScriptAJAXApacheASP', 'MarketingSalesArticle', 'AutoCADBill of', 'DesignT-ShirtsIllustrator', 'ProcessingeCommerceExcelGeneral', 'ModellingCAD/CAMAutoCADSolidworksMechanical', 'AnimationAdobe', 'Photography', 'Agency', 'AdsInstagram', 'QuickBooksMYOB', 'AdsVideo', 'DesignPhotoshopPhoto', 'EditingAfter', 'DesignHTMLJavaScriptLogo', 'EntryExcelLeadsLogo', 'DesignIllustrationAnimationIllustratorPhotoshop', 'HTMLMySQLPHPAPIBackend', 'Active', 'DesignVectorization', 'MarketingAdvertisingGoogle', 'DeveloperApple', 'AnalyticsKeyword', 'Salesforce.comJavaScriptCodeigniterCRMWebsite', 'ProcessingAdvertisement', 'IdentityLogo', 'Job', 'SupportChat', 'DesignBrandingBusiness', '.NETPHPNode.jsHTMLMySQL', 'EntryeCommerceFacebook'

In [91]:
len(unique_skill)

6907

In [102]:
df.drop(columns=['category'], inplace=True)
df.drop(columns=['primary_category'], inplace=True)

In [116]:

def clean_skills(skills):
    cleaned_skills = []
    for skill in skills:
        split_skills = re.findall(r'[A-Z]?[a-z]+(?: [A-Z][a-z]+)*', skill)
        cleaned_skills.extend(split_skills)
    return cleaned_skills

df['skills_list_standardization'] = df['skills_list'].apply(clean_skills)


df.head()

name price     country  \
0  OweBest Technologies Pvt.    18       India   
1     Elite Information Tech    15       India   
2        TAFSOL Technologies    45    Pakistan   
3               eTranslators    30  Bangladesh   
4  ✈THE MARKETING MAVERICKS™    45       India   

                                         description rating reviews Region  \
0  OweBest is a dynamic and rapidly expanding IT ...    4.9     490   Asia   
1  7000 ⭐⭐⭐⭐⭐+ Reviews, 2000+ satisfied clients G...    4.9    7018   Asia   
2  TAFSOL TECHNOLOGIES Group!! Aiming to deliver ...    5.0     195   Asia   
3  eTranslators offers over a DECADE of experienc...    4.9    3633   Asia   
4  The Grace Of Almighty God ____________________...    5.0     558   Asia   

                                         skills_list  skills_count    level  \
0  [PHPWebsite, DesignHTMLMySQLSoftware, Architec...             3  level-5   
1  [SEOLink, BuildingInternet, MarketingAdvertisi...             4  level-5   
2  [PHPWebsite, DesignHTMLSoftware, ArchitectureG...             4  level-5   
3  [TranslationEnglish (US), TranslatorEnglish (U...             4  level-5   
4  [Internet, MarketingFacebook, MarketingSocial,...             5  level-3   

                                       categories  \
0  [Graphic Design, Programming, Web Development]   
1                                     [Marketing]   
2  [Graphic Design, Programming, Web Development]   
3                                   [Translation]   
4                                         [Other]   

                         skills_list_standardization  
0      [Website, Design, My, Software, Architecture]  
1  [Link, Building, Internet, Marketing, Advertis...  
2  [Website, Design, Software, Architecture, Grap...  
3  [Translation, English, Translator, English, Tr...  
4  [Internet, Marketing, Facebook, Marketing, Soc...

In [118]:
unique_skill = list(set(sum(df["skills_list"], [])))
len(unique_skill)

6907

In [117]:
unique_skill = list(set(sum(df["skills_list_standardization"], [])))
len(unique_skill)

1170

In [119]:
standardization_map = {
    "Web Development": ["Website", "Web", "HTML", "CSS", "JavaScript", "PHP"],
    "Graphic Design": ["Design", "Logo", "Illustrator", "Photoshop", "Graphic"],
    "Marketing": ["SEO", "Internet Marketing", "Advertising", "Facebook Marketing", "Social Media"],
    "Programming": ["Python", "Java", "C++", "C#", "Software", "Architecture"],
    "Translation": ["Translation", "Translator", "English", "Spanish", "French"],
    "Content Writing": ["Writing", "Copywriting", "Blog", "Article"],
    "Finance": ["Accounting", "Financial Analysis", "Excel", "Bookkeeping"],
}


def categorize_skills(skills):
    categories = set()  
    for skill in skills:
        for category, keywords in standardization_map.items():
            if skill in keywords:
                categories.add(category)
    return list(categories) if categories else ["Other"]

df['categories'] = df['skills_list_standardization'].apply(categorize_skills)


In [ ]:
df[["categories"]].head(100)

skills_list_standardization  \
0       [Website, Design, My, Software, Architecture]   
1   [Link, Building, Internet, Marketing, Advertis...   
2   [Website, Design, Software, Architecture, Grap...   
3   [Translation, English, Translator, English, Tr...   
4   [Internet, Marketing, Facebook, Marketing, Soc...   
..                                                ...   
95  [Graphic, Design, Website, Design, Logo, Desig...   
96             [Website, Design, Joomla, Word, Press]   
97  [Data, Entry, Virtual, Assistant, Customer, Su...   
98    [Website, Design, Word, Press, Graphic, Design]   
99            [Word, Press, Android, Website, Design]   

                                        categories  
0   [Graphic Design, Programming, Web Development]  
1                                      [Marketing]  
2   [Graphic Design, Programming, Web Development]  
3                                    [Translation]  
4                                          [Other]  
..                                             ...  
95               [Graphic Design, Web Development]  
96               [Graphic Design, Web Development]  
97                                       [Finance]  
98               [Graphic Design, Web Development]  
99               [Graphic Design, Web Development]  

[100 rows x 2 columns]

In [134]:
df["description"]

0       OweBest is a dynamic and rapidly expanding IT ...
1       7000 ⭐⭐⭐⭐⭐+ Reviews, 2000+ satisfied clients G...
2       TAFSOL TECHNOLOGIES Group!! Aiming to deliver ...
3       eTranslators offers over a DECADE of experienc...
4       The Grace Of Almighty God ____________________...
                              ...                        
8456    If you have been scouring the internet looking...
8457    Emdyarts is a highly acclaimed agency on Freel...
8458    My name is Bhawan Baweja, I am a full stack de...
8459    I'm a Senior Web Developer with over 8+ years ...
8460    Live Experts® LLC: Experienced Engineers and D...
Name: description, Length: 8401, dtype: object

In [135]:
df["description"].to_csv("description.csv", index=False)

In [158]:
df['description_Length'] = df['description'].apply(len)

In [142]:
df["description"][0]

'OweBest is a dynamic and rapidly expanding IT company based in India. We assist our customers in identifying their most significant business opportunities and capitalizing on them through our exceptional services. Our expertise includes: 1. Artificial Intelligence (AI) 2. Web Design & Development 3. Mobile Application Development 4. Core PHP Development 5. CodeIgniter, CakePHP, Laravel Development 6. WordPress, Python, Magento Development 7. Node.js, AngularJS, Express.js, React 8. .NET MVC 9. AWS, Azure, and Google Cloud Solutions 10. Hybrid and Native Mobile App Development 11. Android & iOS App Development 12. Salesforce 13. Machine Learning 14. Blockchain Development 15. Flutter and React Native 16. MongoDB, Cassandra, Big Data 17. Cryptocurrency Exchange 18. Penetration & Vulnerability Testing We also work with many other technologies. With OweBest, you can be assured of quality and genuine work. Take a look at our portfolio - it speaks for itself. Best Regards, Team OweBest'

c:\Users\YOUSSEF\AppData\Local\Programs\Python\Python310\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [ ]:
file_path = 'description.csv' 
data = df["description"]


In [150]:
data=pd.DataFrame(data)

In [151]:
data.head()

description
0  OweBest is a dynamic and rapidly expanding IT ...
1  7000 ⭐⭐⭐⭐⭐+ Reviews, 2000+ satisfied clients G...
2  TAFSOL TECHNOLOGIES Group!! Aiming to deliver ...
3  eTranslators offers over a DECADE of experienc...
4  The Grace Of Almighty God ____________________...

In [161]:
vectorizer = CountVectorizer(stop_words='english', max_features=50)  
X = vectorizer.fit_transform(df['description'])
skills = vectorizer.get_feature_names_out()

In [162]:
def extract_skills(description):
    words = description.split()
    extracted_skills = [word for word in words if word.lower() in skills]
    return ', '.join(extracted_skills)

df['Extracted_Skills_from_description'] = df['description'].apply(extract_skills)



In [168]:
df['Extracted_Skills_from_description_list'] = df['Extracted_Skills_from_description'].apply(lambda x: list(set(x.split(','))) if isinstance(x, str) else [])
df['Extracted_Skills_from_description_number'] = df['Extracted_Skills_from_description'].apply(len)


In [170]:
df[['Extracted_Skills_from_description_list', 'Extracted_Skills_from_description_number']].head()

Extracted_Skills_from_description_list  \
0  [ Web,  Development,  quality,  expertise,  So...   
1  [ Data,  Experience,  Years,  social,  experti...   
2  [ Experience,  experience,  Years,  web,  Logo...   
3  [ quality,  MANAGEMENT,  services,  High,  DEV...   
4  [ Web,  business,  services,  Business,  exper...   

   Extracted_Skills_from_description_number  
0                                       221  
1                                       420  
2                                       464  
3                                       158  
4                                       186

In [171]:
output_file_path = 'freelancer_data_after_cleen.csv'
df.to_csv(output_file_path, index=False)


In [172]:
df.columns

Index(['name', 'price', 'country', 'description', 'rating', 'reviews',
       'Region', 'skills_list', 'skills_count', 'level', 'categories',
       'skills_list_standardization', 'description_Length',
       'Extracted_Skills_from_description',
       'Extracted_Skills_from_description_list',
       'Extracted_Skills_from_description_number'],
      dtype='object')